### PR TITLE
Makes it possible to use gpt-3.5

### LINKED ISSUE #12: The model: `gpt-4` does not exist
Hi,
Amazing work you're doing here! ❤
I'm getting this message:

> 
> Traceback (most recent call last):
>   File "/Users/stas/Auto-GPT/AutonomousAI/main.py", line 154, in <module>
>     assistant_reply = chat.chat_with_ai(prompt, user_input, full_message_history, mem.permanent_memory, token_limit)
>                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/Users/stas/Auto-GPT/AutonomousAI/chat.py", line 48, in chat_with_ai
>     response = openai.ChatCompletion.create(
>                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/openai/api_resources/chat_completion.py", line 25, in create
>     return super().create(*args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/openai/api_resources/abstract/engine_api_resource.py", line 153, in create
>     response, _, api_key = requestor.request(
>                            ^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/openai/api_requestor.py", line 226, in request
>     resp, got_stream = self._interpret_response(result, stream)
>                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/openai/api_requestor.py", line 619, in _interpret_response
>     self._interpret_response_line(
>   File "/usr/local/lib/python3.11/site-packages/openai/api_requestor.py", line 679, in _interpret_response_line
>     raise self.handle_error_response(
> **openai.error.InvalidRequestError: The model: `gpt-4` does not exist**

Does it mean I don't have an API access to gpt-4 yet? How can I use previous versions?

### LINKED ISSUE #40: Error: Invalid JSON
I'm getting this argument in a loop

NEXT ACTION:  COMMAND = Error: ARGUMENTS = Invalid JSON
SYSTEM:  Command Error: returned: Unknown command Error:
Error: Invalid JSON

### PR DESCRIPTION
**Pull request by Taytay**
(This is such a neat experiment. Thanks for open-sourcing!)

Fixes #12
Fixes #40 

Makes davinci-3.5-turbo the default model (faster and cheaper)
Allows for model configuration

I made JSON parsing a lot more forgiving, including using GPT to fix up the JSON if necessary.
I also improved the prompt to make it follow instructions a bit better.
I also refactored a bit. 
I also made it save my inputs to a yml file by default when it runs. (This could be a LOT better - very rough, but wanted to get this in)

P.S. Mind adding an MIT license to this?

### CODE REVIEW

**Comment by zorrobyte** [comment]
Tried this, however
```rossfisher@Rosss-MBP scripts % python3 main.py                        
playsound is relying on a python 2 subprocess. Please use `pip3 install PyObjC` if you want playsound to run more efficiently.
Traceback (most recent call last):
  File "/Users/rossfisher/Auto-GPT/scripts/main.py", line 3, in <module>
    import commands as cmd
  File "/Users/rossfisher/Auto-GPT/scripts/commands.py", line 8, in <module>
    import ai_functions as ai
  File "/Users/rossfisher/Auto-GPT/scripts/ai_functions.py", line 11, in <module>
    def call_ai_function(function, args, description, model=cfg.smart_llm_model):
                                                            ^^^^^^^^^^^^^^^^^^^
AttributeError: 'Config' object has no attribute 'smart_llm_model'```

**Comment by Taytay** [comment]
Woah - lemme see what I missed...

**Comment by Taytay** [comment]
Apparently I was moving _way_ too fast. I missed some pretty basic stuff from my commit, and am in the process of fixing it now. (Not at my desk though, so taking me a bit)

**Comment by Taytay** [comment]
Okay - pushed up fixes! Lemme know how that works for you.

**Comment by zorrobyte** [comment]
Seems to work! Need to add my key

May want to upgrade requirements

```Traceback (most recent call last):
  File "/Users/rossfisher/Auto-GPT/scripts/main.py", line 15, in <module>
    from ai_config import AIConfig
  File "/Users/rossfisher/Auto-GPT/scripts/ai_config.py", line 1, in <module>
    import yaml
ModuleNotFoundError: No module named 'yaml'
rossfisher@Rosss-MBP scripts % pip3 install yaml
ERROR: Could not find a version that satisfies the requirement yaml (from versions: none)
ERROR: No matching distribution found for yaml
rossfisher@Rosss-MBP scripts % pip3 install pyyaml
Collecting pyyaml
  Using cached PyYAML-6.0-cp311-cp311-macosx_11_0_arm64.whl (167 kB)
Installing collected packages: pyyaml
Successfully installed pyyaml-6.0```

**Comment by Torantulino** [comment]
Great work, taking a look now.

**Comment by anslex** [comment]
gpt-3.5-turbo
Python 3.11.2
commit https://github.com/Torantulino/Auto-GPT/commit/64af11f4ab9b4fe5e6813a3cbe3d343a1f1b5c98

File is present in 'auto_gpt_workspace'

`python scripts/main.py speak-mode` and error still there:
```
Goal 1: read file 'main.py'
Goal 2: 
Error: Prompt file not found
Error: Invalid JSON
 What is the path or location of the 'main.py' file that you want me to read? Is it in the same directory as this program or is it in a different location?
NEXT ACTION:  COMMAND = Error: ARGUMENTS = Invalid JSON
```

**Comment by Taytay** [comment]
@zorrobyte : Hmmm - I added pyyaml to requirements.txt:
`pip install -r ./requirements.txt` should have picked it up...

**Comment by Torantulino** [comment]
This is amazing!
So much faster and cheaper, excellent work!

**Comment by Taytay** [comment]
I'm fixing merge conflicts now

File: scripts/ai_functions.py
Comment (Torantulino) [inline scripts/ai_functions.py]: Have you found AI_Functions to work on gpt.3.5-turbo?

This is the results of my testing:

| Description               | GPT-4 Result | GPT-3.5-turbo Result | Reason |
|---------------------------|--------------|----------------------|--------|
| Generate fake people      | PASSED       | FAILED               | Incorrect response format |
| Generate Random Password  | FAILED       | FAILED               | Incorrect response format |
| Calculate area of triangle| FAILED       | FAILED               | Incorrect float value (GPT-4), Incorrect response format (GPT-3.5-turbo) |
| Calculate the nth prime number | PASSED  | PASSED               | N/A    |
| Encrypt text              | PASSED       | PASSED               | N/A    |
| Find missing numbers      | PASSED       | PASSED               | N/A    |

File: scripts/chat.py
Comment (Torantulino) [inline scripts/chat.py]: Should this be hard-coded?
Maybe switch this to use the config or user argument

File: scripts/config.py
Comment (Torantulino) [inline scripts/config.py]: I like this idea, but they're currently set to the same thing.

**Comment by Taytay** [comment]
Okay - pushed some fixes. It looks like the bot is inserting its thoughts before the JSON response. I have some ideas to fix, but in the meantime, the JSON parser is doing its job.

I think we need to give it a better example of the response we want. In the meantime, we are seeing stuff like this:
```
assistant reply: As the first step, I need to clone the repository locally so that I can analyze and test the existing code. I will go with the following command:

{
    "command": {
        "name": "execute_python_file",
<snip>...
```

**Comment by Taytay** [comment]
@anslex : Yeah, if you aren't in the scripts folder when you run it, it fails. We'll get that fixed up soon I'm sure.

In the meantime:
```sh
cd ./scripts
python ./main.py speak-mode
```

**Comment by 0xcha05** [comment]
Great work @Taytay !

**Comment by Torantulino** [comment]
@Taytay Eager to merge this, just checking you didn't miss my review comments above! 😊

**Comment by Taytay** [comment]
Which ones actually? I don't see them

**Comment by xSNYPSx** [comment]
Guys, please make some kind of quick switch between models between 3.5 and 4, so that there is one repository, but for example there is a model_config file where we select once and everything else is already contained in the general repository.

**Comment by yousefissa** [comment]
It should be configured in the `.env` file

**Comment by Torantulino** [comment]
Ah sorry @Taytay, new to this, they weren't "sent".

File: scripts/config.py
Comment (Taytay) [reply scripts/config.py]: Yeah, I set them to be the same for now. But I saw you were using GPT-3 for some of the things where you didn't need it to be super smart

File: scripts/ai_functions.py
Comment (Taytay) [reply scripts/ai_functions.py]: Good point - let's get these added to the tests and see if we can get them working with some prompt finagling?

File: scripts/chat.py
Comment (Taytay) [reply scripts/chat.py]: Oh right! Yeah, I need to change that.

File: scripts/ai_functions.py
Comment (Torantulino) [reply scripts/ai_functions.py]: If you like! Submit a pull request over at [AI Functions](https://github.com/Torantulino/AI-Functions).
For now let's make sure we only use 3.5 for AI functions such as json parsing, using your smart/fast model property.

File: scripts/config.py
Comment (Torantulino) [reply scripts/config.py]: Yeah it's a good idea, let's get smart set to gpt-4, then apply them appropriately throughout the code.

**Comment by DataBassGit** [comment]
This worked for me.

The load_prompt function reads the prompt.txt file from the data subdirectory. If the script is not able to access the file, the most likely reason is that the working directory of the script execution is not set correctly.

To resolve this issue, you can modify the load_prompt function to use an absolute path instead of a relative path. You can achieve this by constructing the absolute path using the os module:

```
import os

def load_prompt():
    try:
        # Get the current script's directory
        script_dir = os.path.dirname(os.path.realpath(__file__))
        
        # Construct the absolute path to the prompt.txt file
        prompt_file_path = os.path.join(script_dir, "data", "prompt.txt")

        # Load the prompt from data/prompt.txt
        with open(prompt_file_path, "r") as prompt_file:
            prompt = prompt_file.read()

        return prompt
    except FileNotFoundError:
        print("Error: Prompt file not found", flush=True)
        return ""
```

This modification should fix the issue with accessing the prompt.txt file. The os.path.dirname(os.path.realpath(__file__)) line retrieves the directory of the data.py script, and the os.path.join(script_dir, "data", "prompt.txt") line constructs the absolute path to the prompt.txt file.

File: requirements-alternative.txt
Comment (jcp) [inline requirements-alternative.txt:1]: There has to be a better and cleaner way to approach this. If the environment is isolated properly, it shouldn't matter.

**Comment by Torantulino** [comment]
Testing now 👀
Let's get this shipped! 🚢

**How to test this on your machine:**
`git checkout -b Taytay-fixes_gpt3 master`
`git pull https://github.com/Taytay/Auto-GPT.git fixes_gpt3`

**Comment by Torantulino** [comment]
Please remove all that debug logging (or make it optional) before we merge this @Taytay 😊

File: scripts/chat.py
Comment (Taytay) [reply scripts/chat.py]: Fixed!

File: scripts/config.py
Comment (Taytay) [reply scripts/config.py]: I did my best here. Reserved the smart LLM for some of the more complicated stuff.

File: requirements-alternative.txt
Comment (Taytay) [reply requirements-alternative.txt:1]: I totally agree - I'm just trying to get something working. :)
By default, everyone can use the requirements.txt. Me and few other folks can't for some reason. (On WSL or Mac, I get this same error)

I'm in a totally new conda environment, and when I use the existing requirements.txt, it fails with this error:
```
# The conflict is caused by:
#     The user requested requests>=2.26.0
#     docker 6.0.1 depends on requests>=2.26.0
#     googlesearch-python 1.1.0 depends on requests==2.25.1
```
It looks like a funny dependency between `googlesearch-python` and `docker`. 

I'm happy to solve it a better way. I'm just too new to python. :)

File: scripts/config.py
Comment (Torantulino) [inline scripts/config.py:38]:
```suggestion
        # Initialize the OpenAI API client
        openai.api_key = self.openai_api_key
```

File: scripts/config.py
Comment (Torantulino) [inline scripts/config.py]:
```suggestion
import openai
```

File: requirements-alternative.txt
Comment (jcp) [reply requirements-alternative.txt:1]: I'm on WSL2 too. I'll try to reproduce this in the morning. I haven't had issues but I've opted to isolate my environment other ways that aren't in the docs.

This relates to my discussion here: https://github.com/Torantulino/Auto-GPT/discussions/48.

**Comment by Taytay** [comment]
@Torantulino - lemme know how that works for you. I think it's in better shape now.

**Comment by Taytay** [comment]
I made this use gpt-4 again for the main prompt again. (Not sure if that's what you wanted or not.) I think we can make it use gpt-3 with some better prompting.

**Comment by yousefissa** [comment]
heads up, i've been testing this branch and often run into: `openai.error.InvalidRequestError: This model's maximum context length is 4097 tokens. However, your messages resulted in 4304 tokens. Please reduce the length of the messages.`

or JSON decode errors along the way

**Comment by Taytay** [comment]
@yousefissa : You using gpt-4 as your "smart" llm, or gpt-3?

**Comment by Taytay** [comment]
Any chance you could show me some of the JSON it's not parsing?

**Comment by yousefissa** [comment]
i don't have gpt-4 api access yet so using 3.5 for both llm options.


here is a larger rip of the log output:

```
System: Command browse_website returned: Website Content Summary:  "Result" : The subpage promotes Shopify's features and solutions for e-commerce businesses, including website creation, app integration, selling and marketing tools, and enterprise-level retail solutions. It also offers a free 3-day trial and a point of sale system for physical stores. The focus is on the built-in marketing tools and centralized business operations available with a Shopify POS, and the subpage includes testimonials and resources for support and strategy tips.

Links: ['Skip to Content (#main)', 'Sign up now (https://accounts.shopify.com/store-create?signup_currency=USD&locale=en&language=en&signup_page=https%3A%2F%2Fwww.shopify.com%2F&signup_types[]=paid_trial_experience)', 'Shopify (https://www.shopify.com)', 'Start your businessBuild your brand (/start)', 'Create your websiteOnline store editor (/website/builder)']
User: Determine which next command to use, and respond using the format specified above:
Assistant: {
    "command": {
        "name": "memory_add",
        "args":{
            "string": "Researched in-depth information on Shopify on their website."
        }
    },
    "thoughts":
    {
        "text": "Adding Shopify research task to permanent memory.",
        "reasoning": "Since Shopify is the platform we are planning to use for our business needs, it's necessary to conduct further research to fully assess its capabilities to ensure our store remains profitable in the near future.",
        "plan": "- Store info on Shopify on memory",
        "criticism": "None",
        "speak": "I have conducted thorough research on Shopify and have added relevant info to be used in decision-making to permanent memory."
    }
}
System: Command memory_add returned: Committing memory with string "Researched in-depth information on Shopify on their website."
User: Determine which next command to use, and respond using the format specified above:
Assistant: {
    "command": {
        "name": "message_agent",
        "args":{
            "key": "-1",
            "message": "Is it possible to add an email newsletter sign-up form to my Shopify store?"
        }
    },
    "thoughts":
    {
        "text": "Message the GPT-3.5-powered Shopify Help Agent for information on integrating an email newsletter sign-up form.",
        "reasoning": "Asking the Shopify help agent for relevant information on the email newsletter sign-up form can provide easy-to-understand guidance on the subject. This way, we can use their suggested implementation method to add an email newsletter sign-up form to our store.",
        "plan": "- Message the Shopify help agent on Shopify integration methods for email newsletters.",
        "criticism": "None",
        "speak": "I have messaged the Shopify Help Agent for information on integrating an email newsletter sign-up form."
    }
}
System: Command message_agent returned: Error: -1
User: Determine which next command to use, and respond using the format specified above:
Assistant: {
    "command": {
        "name": "list_agents",
        "args":{
        }
    },
    "thoughts":
    {
        "text": "Listing all available agent.",
        "reasoning": "It seems I have sent a message to a non-existing agent. To avoid future errors, I will list all available agents and message the correct one.",
        "plan": "- List available agents.",
        "criticism": "None",
        "speak": "I will list all available agents to message the correct one."
    }
}
System: Command list_agents returned: []
User: Determine which next command to use, and respond using the format specified above:
----------- END OF CONTEXT ----------------
Traceback (most recent call last):
  File "/Users/me/Desktop/ai-projects/Auto-GPT/scripts/main.py", line 284, in <module>
    assistant_reply = chat.chat_with_ai(
  File "/Users/me/Desktop/ai-projects/Auto-GPT/scripts/chat.py", line 65, in chat_with_ai
    assistant_reply = create_chat_completion(
  File "/Users/me/Desktop/ai-projects/Auto-GPT/scripts/llm_utils.py", line 9, in create_chat_completion
    response = openai.ChatCompletion.create(
  File "/Users/me/Desktop/ai-projects/Auto-GPT/venv/lib/python3.10/site-packages/openai/api_resources/chat_completion.py", line 25, in create
    return super().create(*args, **kwargs)
  File "/Users/me/Desktop/ai-projects/Auto-GPT/venv/lib/python3.10/site-packages/openai/api_resources/abstract/engine_api_resource.py", line 153, in create
    response, _, api_key = requestor.request(
  File "/Users/me/Desktop/ai-projects/Auto-GPT/venv/lib/python3.10/site-packages/openai/api_requestor.py", line 226, in request
    resp, got_stream = self._interpret_response(result, stream)
  File "/Users/me/Desktop/ai-projects/Auto-GPT/venv/lib/python3.10/site-packages/openai/api_requestor.py", line 619, in _interpret_response
    self._interpret_response_line(
  File "/Users/me/Desktop/ai-projects/Auto-GPT/venv/lib/python3.10/site-packages/openai/api_requestor.py", line 682, in _interpret_response_line
    raise self.handle_error_response(
openai.error.InvalidRequestError: This model's maximum context length is 4097 tokens. However, your messages resulted in 4117 tokens. Please reduce the length of the messages.
(venv) ➜  scripts git:(fixes352) ✗
```


ill try to log JSON output next time

**Comment by Taytay** [comment]
In that case, try setting the context to be much smaller in config.py. Set it to 2000 instead and see how that does.
It's a bit step down, but gpt-4 has an 8k windows, and gpt-3, a 4k window, so 2000 is probably a better default for gpt-3

**Comment by yousefissa** [comment]
I'll give that a go, thanks!

**Comment by Torantulino** [comment]
@Taytay, why are we now seeing `AttributedDict([(` being passed as commands?

image

**Comment by jcp** [comment]
@Taytay, perhaps these can be env variables too with defaults? Assuming I'm following what the issue is. I may not be. 4,000 seems high for gpt-3 though. 

https://github.com/Torantulino/Auto-GPT/pull/45/files#diff-e9ffa9c36082d5a0b2ae4098b1b77d020a8d57d159bc89255fcad17287bc3ed9R36-R39

Context windows are frustrating. I wish there was a more eloquent way to optimize limits.

**Comment by yousefissa** [comment]
I also occasionally see `SYSTEM:  Command google returned: Error: module 'browse' has no attribute 'search'`

i believe this is because we attempt to use `browse.search`. we should probably update to:

`from googlesearch import search` in `commands.py`

**Comment by Torantulino** [comment]
Your call to the `fix_json` function doesn't include a schema! 
Your code supports it, but you're passing None, is this intentional?

@Taytay 
image

**Comment by Taytay** [comment]
> @Taytay, why are we now seeing `AttributedDict([(` being passed as commands?
> 
> 
> 
> image
> 
> 

Good question - I wonder if it's the dirty parse library? I started seeing that so quickly after using it that I didn't debug properly.

**Comment by Taytay** [comment]
> @Taytay, perhaps these can be env variables too with defaults? Assuming I'm following what the issue is. I may not be. 4,000 seems high for gpt-3 though. 
> 
> 
> 
> https://github.com/Torantulino/Auto-GPT/pull/45/files#diff-e9ffa9c36082d5a0b2ae4098b1b77d020a8d57d159bc89255fcad17287bc3ed9R36-R39
> 
> 
> 
> Context windows are frustrating. I wish there was a more eloquent way to optimize limits. 

I think that's a good idea too. What I really want is a way to pre-create the various llm's with their associated config at startup, rather than pass strings/into around. But if we do that, we should probably just use langchain's abstractions and initialize instances of their LLM class.

**Comment by Taytay** [comment]
I won't be able to do much more on this for another day or so, if folks want to try to take this over the finish line.

It's been fun hacking with you guys!

**Comment by Torantulino** [comment]
Thank you so much for all your hard work @Taytay, we'll take it from here!
Due to popular request I'll be launching a discord soon, so there will be a dedicated space to chat and develop faster.

**Comment by jcp** [comment]
Very true. Langchain integration would cut out a good chunk of code. 

I also have my eye on this: https://github.com/Torantulino/Auto-GPT/blob/8cf4bf9278e09ad27b1ff81c3622fce2abbee385/scripts/commands.py#L14-L92. This logic could be handled by the model itself alongside a plugin register. I did something similar a week ago in a Colab and it worked out really well.

**Comment by DataBassGit** [comment]
This script did resolve our json error:

https://github.com/Torantulino/Auto-GPT/pull/45#issuecomment-1493533003

We found that it was from the prompt not being able to be reached by data.py.

Also I see you offloaded call_ai_function to an external library like I suggested. Tomorrow, we're going to try to test hooking this up to discord.

**Comment by Torantulino** [comment]
Multiple fixes inbound.

- [ ] Maximum context length reached Error
- [ ] Broken Google Search Command
- [ ] Vastly Improved JSON parsing for gpt3.5

**Review by Torantulino — APPROVED** [review]

**Comment by SirGunthix** [comment]
what do we need to do to fix the JSON Error? I am still getting it. Do I need to do anything different than what's in the readme? If so, can we update readme to include details? Or you can just tell me here and Ill try to push a new cmomit (this will be my first commit on github ever :)

**Comment by Apollyon81** [comment]
> @anslex: Sim, se você não estiver na pasta de scripts ao executá-lo, ele falhará. Vamos consertar isso em breve, tenho certeza.
> 
> Enquanto isso:
> 
> ```shell
> cd ./scripts
> python ./main.py speak-mode
> ```
how solve this??
python ./main.py speak-mode
python: can't open file '/workspaces/Auto-GPT-0.2.2/scripts/./main.py': [Errno 2] No such file or directory

**Comment by caffo** [comment]
@Torantulino, any idea when we will see this on Stable (the docker image in special)?

**Comment by Torantulino** [comment]
Sorry @caffo, I'm not sure what it is you're reffering to, can you elaborate?

**Comment by caffo** [comment]
@Torantulino Sorry, I think I got confused. This is already on the stable docker image as the flag --gpt3-only, right?

**Comment by Torantulino** [comment]
Ah I see, yes this issue was actually to make GPT3 _possible_ to use at all in the first place, because at the start of this project it only worked with gpt4.

You are correct that gpt3only is available to use, and the current default is a mix where gpt3 is used the vast majority of the time.

**Pull request merged**
